pyvelox/complex.cpp

```diff
@@ -0,0 +1,160 @@
+  }
+}
+
+template <TypeKind Kind>
+void setElementInFlatVector(
+    vector_size_t idx,
+    const variant& v,
+    VectorPtr& vector) {
+  using NativeType = typename TypeTraits<Kind>::NativeType;
+  auto asFlat = vector->asFlatVector<NativeType>();
+  asFlat->set(idx, NativeType{v.value<NativeType>()});
+}
+
+void constructType(TypePtr& type, const variant& v, ElementCounter& counter) {
+  ++counter.totalElements;
+
+  if (v.isNull()) {
+    // since the variant is NULL, we can't infer the data type
+    // of it's infer, thus it maybe UNKNOWN or INVALID at this stage
+    // which implies further investigation is required
+    if (v.kind() != TypeKind::UNKNOWN && v.kind() != TypeKind::INVALID &&
+        v.kind() != type->kind()) {
+      throw std::invalid_argument("Variant was of an unexpected kind");
+    }
+    return;
+  } else {
+    // if a Non-Null variant's type is unknown or not one of the valid
+    // types which are supported then the Type tree cannot be constructed
+    if (v.kind() == TypeKind::UNKNOWN || v.kind() == TypeKind::INVALID) {
+      throw std::invalid_argument(
+          "Non-null variant has unknown or invalid kind");
+    }
+
+    switch (v.kind()) {
+      case TypeKind::ARRAY: {
+        counter.children.resize(1);
+        auto asArray = v.array();
+        TypePtr childType = createType(TypeKind::UNKNOWN, {});
+        for (const auto& element : asArray) {
+          constructType(childType, element, counter.children[0]);
+        }
+
+        // if child's type still remains Unknown, implies all the
+        // elements in the array are actually NULL
+        if (childType->kind() == TypeKind::UNKNOWN) {
+          throw py::value_error("Cannot construct array with all None values");
+        }
+        checkOrAssignType(type, [&childType]() {
+          return createType<TypeKind::ARRAY>({childType});
+        });
+        break;
+      }
+
+      default: {
+        checkOrAssignType(type, [&v]() { return createScalarType(v.kind()); });
+        break;
+      }
+    }
+  }
+}
+
+static void insertVariantIntoVector(
+    TypeKind& typeKind,
+    const variant& v,
+    VectorPtr& vector,
+    ElementCounter& counter) {
+  if (v.isNull()) {
+    vector->setNull(counter.insertedElements, true);
+  } else {
+    switch (typeKind) {
+      case TypeKind::ARRAY: {
+        auto asArray = vector->as<ArrayVector>();
+        asArray->elements()->resize(counter.children[0].totalElements);
+        const std::vector<variant>& elements = v.array();
+        vector_size_t offset = 0;
+        if (counter.insertedElements != 0) {
+          offset = asArray->offsetAt(counter.insertedElements - 1) +
+              asArray->sizeAt(counter.insertedElements - 1);
+        }
+        asArray->setOffsetAndSize(
+            counter.insertedElements, offset, elements.size());
+        for (const variant& elt : elements) {
+          auto elt_type = elt.kind();
```

Comment:
nit: lets stick to velox convention so eltType not elt_type. Also call it eltKind